[Concept Entry] Rust Generics #3878 (#3984)
Codecademy/docs: content/rust/concepts/generics/generics.md

---
Title: 'Generics'
Description: 'Generics are a programming feature that allows the programmer to write versatile and reusable code that operates on different data types.'
Subjects:
  - 'Computer Science'
  - 'Game Development'
Tags:
  - 'Generics'
  - 'Data Types'
CatalogContent:
  - 'rust-for-programmers'
  - 'paths/computer-science'
---

> In Rust, **Generics** are tools that enhance code flexibility and reusability. They enable functions, structs, enums, and traits to work with different data types while ensuring type safety. Generics help in reducing code duplication, increasing efficiency, and allowing for abstract and scalable code designs.

Reviewer: ```suggestion
In Rust, **Generics** are tools that enhance code flexibility and reusability. They enable functions, structs, enums, and traits to work with different data types while ensuring type safety. Generics help reduce code duplication, increase efficiency, and allow abstract and scalable code designs.
```

## Implementation

### Functions

Generics enable functions to handle arguments of various types, promoting code reuse and flexibility. In the following example, `generic_function`, defined with `<T>`, illustrates how a single function can adapt to various argument types.

```

> fn generic_function<T>(item: T) {
>     println!("{}", item);
> }
> generic_function(1);                 // prints: 1
> generic_function("Hello, World!");   // prints: Hello, World!
> generic_function(true);              // prints: true

Reviewer: ```suggestion
fn generic_function<T> (item: T) -> T {
    return item;
}
fn main() {
    println!("{}", generic_function(1));                 // prints: 1
    println!("{}", generic_function("Hello, World!"));   // prints: Hello, World!
    println!("{}", generic_function(true));              // prints: true   
}
```
Now this code block gives the expected output.

```

`<T>` represents a generic type, meaning `generic_function` can accept any type of argument. The `println!("{}", item);` line prints the argument. Since the function is generic, `item` can be of different types, such as an integer, a string, or a boolean.

### Structs

Generics in structs allow for flexible structures capable of storing various types. In the following example, `GenericStruct` uses of generics to create flexible and reusable struct definitions.

```

> struct GenericStruct<T> {
>     firstWord: T,
>     secondWord: T,
> }
> let words_struct = GenericStruct { firstWord: "Hello", secondWord: "World" };

Reviewer: ```suggestion
struct GenericStruct<T> {
    firstWord: T,
    secondWord: T,
}

fn main() {
    let words_struct = GenericStruct { firstWord: "Hello", secondWord: "World" };
    println!("{}, {}", words_struct.firstWord, words_struct.secondWord);       //prints Hello, World
}
```

### Traits

Generics in traits allow for defining shared behavior across different types. In the following example, the `GenericTrait` trait, applied to `ExampleStruct`, shows polymorphism through generic methods.

```
trait GenericTrait<T> {
    fn generic_function(&self, item: T) -> T;
}
struct ExampleStruct<T> {
    value: T,
}
impl<T> GenericTrait<T> for ExampleStruct<T> {
    fn generic_function(&self, item: T) -> T {
        item
    }
}
let example = ExampleStruct { value: 10 };
let result = example.generic_function(20);
println!("Result: {}", result); // Prints: Result: 20
```

In this example, `GenericTrait<T>` defines a generic method `generic_function` which will return the item it is given. `ExampleStruct<T>` implements `GenericTrait<T>`, meaning it adopts the behavior defined in `GenericTrait`. In the last line, `result` calls the `generic_function` method on `example` with an integer argument of `20`, returning `20`.

### Trait Bounds

> Trait bounds are used to specify constraints on generic types, ensuring they adhere to certain behaviors. In the following example, `print_debug` function uses trait bounds to ensure generic types can be safely and predictably debugged.

Reviewer: ```suggestion
Trait bounds are used to specify constraints on generic types, ensuring they adhere to certain behaviors. In the following example, the `print_debug` function uses trait bounds to ensure generic types can be safely and predictably debugged.
```
